ttyplot.c

```diff
@@ -289,9 +290,99 @@ void redraw_screen(const char * errstr) {
     refresh();
 }
 
+// Handle a single value from the input stream.
+// Return whether we got a full data record.
+int handle_value(double value)
+{
+    static double saved_value;
+    static int saved_value_valid = 0;
+
+    // First value of a 2-value record: save it for later.
+    if (two && !saved_value_valid) {
+        saved_value = value;
+        saved_value_valid = 1;
+        return 0;
+    }
+
+    // Otherwise we have a full record.
+    n = (n+1) % plotwidth;
+    if (two) {
+        values1[n] = saved_value;
+        values2[n] = value;
+        saved_value_valid = 0;
+    } else {
+        values1[n] = value;
+    }
+    if (rate)
+        td = derivative(&values1[n], two ? &values2[n] : NULL, &now);
+    return 1;
+}
+
+// Parse the input data. Return the number of bytes consumed.
+size_t parse_input(char *buffer, size_t length)
+{
+    static const char delimiters[] = " \t\r\n";  // white space
+
+    // Find the last delimiter.
+    int end;
+    for (end = length - 1; end >= 0; end--) {
+        if (strchr(delimiters, buffer[end]))  // found
+            break;
+    }
+    if (end < 0)  // no delimiter
+        return 0;
+    buffer[end] = '\0';
+
+    // Tokenize and parse.
+    int records = 0;  // number or records found
+    char *str = buffer;
+    char *token;
+    while ((token = strtok(str, delimiters)) != NULL) {
+        records += handle_value(strtod(token, NULL));
```

Comment:
I have yet to verify that but `strtod(token, NULL)` (with empaphis on `NULL`) should tolerate some garbage around the double value and hence render half-broken input as healthy whereas `sscanf` would have discarded the token altogether. Is that change in behavior intended?

It maybe worth noting here that replacing all-garbage tokens by 0.0 is not a side effect but our explicitly wanted behavior. To a reader not knowing, the code as is looks like we're just forgetting to check for errors. Maybe that's worth a small comment in code.

Reply:
> `strtod()` tolerates suffix garbage, but not prefix garbage. For instance, "x=42" would be refused, whereas "318mV, 35.4°C" would be parsed as {318, 35.4}. I think this is a good thing, as it could alleviate the need for preprocessing upstream.

@edgar-bonet true, but once we start accepting half-garbage there will be no way going back without breaking existing usage. How about we remain strict for now and add a command line argument for a tolerant mode later (not in the this pull request) if we find its needed? I would vote for being strict by default, for now. What do you think?

> > It maybe worth noting here that replacing all-garbage tokens by 0.0 is not a side effect but our explicitly wanted behavior.
> 
> Indeed. But maybe it would be better to just skip garbage tokens? It could make preprocessing still easier for the user. What do you think?

There's a chance that I'm misunderstanding you here, let's see.
Turning garbage into 0.0 seems needed to not mix-up channels with `-2` mode or input "1 garbage 3 garbage 5 6" would be rendered as "<1 3> <5 6>" rather than current "<1 0> <3 0> <5 6>". So skipping seems to be trouble.
Did you have something else in mind?